### PR TITLE
Bug 1891921 - Let mozilla-version handle whether a version is a major, stability, or a development one. 

### PR DESCRIPTION
**Pull request by JohanLorenzo**

### CODE REVIEW

File: mozilla_version/gecko.py [hunk 1]
```diff
@@ -157,6 +157,15 @@ def __attrs_post_init__(self):
...
                 'Beta number and patch number cannot be both defined',
+            ), (
+                self.is_major and self.is_stability,
+                'Version cannot be both a major and a stability one',
+            ), (
+                self.is_major and self.is_development,
+                'Version cannot be both a major and a development one',
+            ), (
+                self.is_stability and self.is_development,
+                'Version cannot be both a stability and a development one',
```
Comment (JohanLorenzo) [inline mozilla_version/gecko.py:168]: Based on the current implementation of these 3 new properties, this errors should never happen. That said, they just helped me get the right implementation. Moreover, they will be useful any time someone changes or subclasses these 3 properties.

File: mozilla_version/gecko.py [hunk 2]
```diff
@@ -528,6 +593,39 @@ class FirefoxVersion(_VersionWithEdgeCases):
...
 
+    _RELEASED_EDGE_CASES_MAJOR = ({
+        'major_number': 1,
+        'minor_number': 5,
+    }, {
+        'major_number': 3,
+        'minor_number': 5,
+    }, {
+        'major_number': 3,
+        'minor_number': 6,
```
Comment (JohanLorenzo) [inline mozilla_version/gecko.py:604]: Version before the introduction of the rapid release scheme didn't have build numbers. Instead, they had release candidates ones.

File: mozilla_version/gecko.py [hunk 3]
```diff
@@ -565,6 +663,23 @@ class FennecVersion(_VersionWithEdgeCases):
         'build_number': 1,
     })
 
+    _RELEASED_EDGE_CASES_MAJOR = ({
+        'major_number': 1,
+        'minor_number': 1,
+    }, {
+        'major_number': 14,
+        'minor_number': 0,
+        'patch_number': 1,
```
Comment (JohanLorenzo) [inline mozilla_version/gecko.py:672]: For an unknown reason, the current product details doesn't any track record of what build number we had on Fennec 14.0.1. That's the reason why it's missing here.

File: mozilla_version/gecko.py [hunk 4]
```diff
@@ -528,6 +593,39 @@ class FirefoxVersion(_VersionWithEdgeCases):
...
+        'major_number': 3,
+        'minor_number': 5,
+    }, {
+        'major_number': 3,
+        'minor_number': 6,
+    }, {
+        'major_number': 14,
+        'minor_number': 0,
+        'patch_number': 1,
+        'build_number': 1,
```
Comment (JohanLorenzo) [inline mozilla_version/gecko.py:609]: Fun fact: product-details currently considers 14.0.1 to be both a stability version[1] and a major one[2]. This implementation will fix this and make it only a major one: 

[1] https://github.com/mozilla-releng/product-details/blob/5b34c7d73075fe31641a0c846cc191d0fa1f004a/public/1.0/firefox_history_stability_releases.json#L120
[2] https://github.com/mozilla-releng/product-details/blob/5b34c7d73075fe31641a0c846cc191d0fa1f004a/public/1.0/firefox_history_major_releases.json#L18

File: mozilla_version/test/test_gecko.py [hunk 5]
```diff
@@ -742,3 +742,1141 @@ def test_gecko_version_hashable():
...
+    "33.1",
+    "33.1-build1",
+    "33.1-build2",
+    "33.1-build3",
+    "34.0",
+    "124.0",
+    "125.0.1",
+    "125.0.1-build1",
+))
+def test_firefox_version_are_major(version_string):
```
Comment (JohanLorenzo) [inline mozilla_version/test/test_gecko.py:769]: Note to future readers: we don't have tests to check whether a version is not a major one. That said, it's already covered by the error reporting GeckoVersion.__attrs_post_init__(). Thus, a version wrongly considered as a major one will raise an error and make other tests fail.

File: mozilla_version/test/test_gecko.py [hunk 6]
```diff
@@ -742,3 +742,1141 @@ def test_gecko_version_hashable():
...
+    "38.0.1-build1",
+    "45.0",
+    "115.0",
+))
+def test_thunderbird_version_are_major(version_string):
+    assert ThunderbirdVersion.parse(version_string).is_major
+
+
+@pytest.mark.parametrize('version_string', (
+    "1.0.1",
```
Comment (JohanLorenzo) [inline mozilla_version/test/test_gecko.py:805]: Note to future readers: Unlike the version above where I singled out edge cases, here I just dumped what product-details had. It's not great but it saved time and I'm sure all existing cases didn't regress.
